### ISSUE TITLE
[YSQL] Colocated range index upper-bound scan fails

### ISSUE THREAD

**Issue by jaki** [yugabyte/yugabyte-db#12899]
Jira Link: [DB-2643](https://yugabyte.atlassian.net/browse/DB-2643)
### Description

Range-partitioned colocated index reverse scan _after TRUNCATE_ fails with DocDB error `Cannot decode value type kGroupEnd from the key encoding format: "!"` (see issues with same message: #10279, #12422).  The truncate is necessary to reproduce.  Repro on recent master commits 25ce96c6bbe8bec5036f95f687e3f5672a12b044 and b661d0a0a71302b3c57097db402f90a6ae11b938:

```sql
CREATE TABLEGROUP g;
CREATE TABLE t (i int) TABLEGROUP g;
CREATE INDEX ON t (i);
TRUNCATE t;
INSERT INTO t VALUES (1), (2), (3);
EXPLAIN SELECT i FROM t WHERE i = 2;
SELECT i FROM t WHERE i = 2;
EXPLAIN SELECT i FROM t WHERE i > 1;
SELECT i FROM t WHERE i > 1;
EXPLAIN SELECT i FROM t WHERE i < 3;
SELECT i FROM t WHERE i < 3;
```

```
ERROR:  Query error: Cannot decode value type kGroupEnd from the key encoding format: "!"
```

For colocated index, default partitioning is ASC.  If the index is specified DESC (i.e. `CREATE INDEX ON t (i DESC)`), the `i > 1` query fails instead of `i < 3`.

Recently, having index separated from table was disallowed (commit 307cf89ab91b50ae060ffb798a1c5835db660b43).  Reverting that restriction, this still fails when table is not colocated but index is.

I noticed that the table tombstone created by colocated truncate still exists after `yb-admin compact_table` for both the table and index.  That is strange when related tests in `src/yb/docdb/docdb-test.cc` pass.  See also related issue #12898.

**Comment by jaki** [comment]
This fortunately appears to not be a permanent storage issue.  After compaction gets rid of all table tombstones, the query works.  Tested using tserver flag `--timestamp_history_retention_interval_sec=30`.

Also tested on system catalog truncate (using in-progress REINDEX work) and found that this issue doesn't occur for that.  Verified with `sst_dump` that table tombstone does appear and also disappears after compaction, just like with user colocated tables.  Specifically tested `pg_depend_reference_index` and `SELECT * FROM pg_depend WHERE refclassid < 1250 AND refobjid < 1000 AND refobjsubid < 1000;` and `SELECT * FROM pg_depend WHERE refclassid < 1250;`.

**Comment by tverona1** [comment]
Another repro:
```
-- create colocated database
CREATE TABLE IF NOT EXISTS t1(c0 boolean);
CREATE TABLE IF NOT EXISTS t2(c0 money  NOT NULL UNIQUE);
INSERT INTO t1(c0) VALUES (true), (FALSE);
TRUNCATE t2;
SELECT SUM(count) FROM (SELECT ALL (t1.c0)::INT as count FROM ONLY t2 RIGHT OUTER JOIN ONLY t1 ON ((t2.c0)<(CAST(0.27233025 AS MONEY)))) as res;
ERROR:  Query error: Cannot decode value type kGroupEnd from the key encoding format: "!"
```

**Comment by rthallamko3** [comment]
@Huqicheng to pick it up as time permits.

**Comment by jasonyb** [comment]
Resolved by recent commit 3d9d8bae563b13010831a5c45ce44b4070126eca.
